flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/utils/ExtractionUtils.java

```diff
@@ -0,0 +1,565 @@
+					return Modifier.isPublic(m) && !Modifier.isNative(m) && !Modifier.isAbstract(m);
+				})
+				.forEach(methods::add);
+			clazz = clazz.getSuperclass();
+		}
+		return methods;
+	}
+
+	// --------------------------------------------------------------------------------------------
+	// Assignable Constructor Utilities
+	// --------------------------------------------------------------------------------------------
+
+	/**
+	 * Result of the extraction in {@link #extractAssigningConstructor(Class, List)}.
+	 */
+	public static class AssigningConstructor {
+		public final Constructor<?> constructor;
+		public final List<String> parameterNames;
+
+		private AssigningConstructor(Constructor<?> constructor, List<String> parameterNames) {
+			this.constructor = constructor;
+			this.parameterNames = parameterNames;
+		}
+	}
+
+	/**
+	 * Checks whether the given constructor takes all of the given fields with matching (possibly
+	 * primitive) type and name. An assigning constructor can define the order of fields.
+	 */
+	public static @Nullable AssigningConstructor extractAssigningConstructor(
+			Class<?> clazz,
+			List<Field> fields) {
+		AssigningConstructor foundConstructor = null;
+		for (Constructor<?> constructor : clazz.getDeclaredConstructors()) {
+			if (!Modifier.isPublic(constructor.getModifiers())) {
+				continue;
+			}
+			final List<String> parameterNames = extractConstructorParameterNames(clazz, constructor, fields);
+			if (parameterNames != null) {
+				if (foundConstructor != null) {
+					throw extractionError(
+						"Multiple constructors found that assign all fields for class '%s'.",
+						clazz.getName());
+				}
+				foundConstructor = new AssigningConstructor(constructor, parameterNames);
+			}
+		}
+		return foundConstructor;
+	}
+
+	/**
+	 * Extracts ordered parameter names from a constructor that takes all of the given fields with
+	 * matching (possibly primitive) type and name.
+	 */
+	private static @Nullable List<String> extractConstructorParameterNames(
+			Class<?> clazz,
+			Constructor<?> constructor,
+			List<Field> fields) {
+		final Type[] parameterTypes = constructor.getGenericParameterTypes();
+		List<String> parameterNames = Stream.of(constructor.getParameters())
+			.map(Parameter::getName)
+			.collect(Collectors.toList());
+
+		// by default parameter names are "arg0, arg1, arg2, ..." if compiler flag is not set
+		// so we need to extract them manually if possible
+		if (parameterNames.stream().allMatch(n -> n.startsWith("arg"))) {
+			final ParameterExtractor extractor = new ParameterExtractor(constructor);
+			getClassReader(clazz).accept(extractor, 0);
+
+			final List<String> extractedNames = extractor.getParameterNames();
+			if (extractedNames.size() == 0 || !extractedNames.get(0).equals("this")) {
+				return null;
+			}
+			// remove "this" and additional local variables
+			parameterNames = extractedNames.subList(1, Math.min(fields.size() + 1, extractedNames.size()));
```

Comment:
Is the removing trailing names correct? Does it mean that we accept a ctor that has more arguments than just the fields?

Reply:
Very good point. This is a bug.